Add some extra functionality.

lrhn opened the pull request:
Also do some clean-up.

Fixes dart-lang/core#378, dart-lang/core#379 and dart-lang/core#381:

* https://github.com/dart-lang/core/issues/381
* https://github.com/dart-lang/characters/issues/22
* https://github.com/dart-lang/core/issues/378

lrhn committed:
Add some extra functionality.

Also do some clean-up.

Fixes #21, #22 and #24:

* https://github.com/dart-lang/characters/issues/24
* https://github.com/dart-lang/characters/issues/22
* https://github.com/dart-lang/characters/issues/21

jakemac53 commented on CHANGELOG.md:
```diff
@@ -1,8 +1,15 @@
+## 1.2.0-nullsafety
```
There hasn't been a stable `1.1.0` release yet - I would just do `1.1.0-nullsafety.3`

lrhn replied:
Done.

lrhn committed:
Format

jakemac53 commented on lib/src/characters_impl.dart:
```diff
@@ -186,60 +188,96 @@ class StringCharacters extends Iterable<String> implements Characters {
+  Characters characterAt(int position) {
+    var breaks = Breaks(string, 0, string.length, stateSoTNoBreak);
+    var start = 0;
+    findIndices:
```
I am personally not a fan of using labels in dart - they are just really uncommon and thus can be confusing or require additional research to understand exactly what they do (I had no idea you could even use them to break out from a whole block like this).

Consider just inlining the two throw statements instead of doing `break findIndices;` which really just throws - or call a method that throws. That has a lot more straightforward/explicit behavior imo.

lrhn replied:
With that attitude, there won't ever be common use of labels 😉 .

This code is probably just as good with inlined throws. Will do.

FYI: You can label any *statement*, and you can break any surrounding label. It only really makes sense for composite statements (but you *can* write `foo: break foo;`). In that way, Dart precisely copied Java and JavaScript, so it's not anything fancy. Since Dart does not have loop `else` branches, like Python, I sometimes write what should be `for (...) { ... if (...) break; ... } else { ... else code ... }` as:
```dart
outer: {
  for (...) {
     ...
     if (...) break outer; 
     ...
  }
  ... else code ...
}
```

lrhn commented on lib/src/characters_impl.dart:
```diff
@@ -186,60 +188,96 @@ class StringCharacters extends Iterable<String> implements Characters {
-      return StringCharacters(string.substring(0, endIndex));
+      var end = breaks.nextBreak();
+      if (end < 0) break findIndices;
+      ;
```
That `;` looks unnecessary :)

lrhn committed:
Address comments.

jakemac53 reviewed (approved):
I think `dartfmt` changed a bit so it looks like the example needs to be formatted again to fix travis, otherwise lgtm

lrhn committed:
Update SDK version to use new formatter.

lrhn committed:
Fix typo

lrhn committed:
Try a later SDK version.

Travis won't run tests with the current set-up, the source span gets a version which doesn't work.
Try a later SDK version to see if that helps.

jakemac53 commented on pubspec.yaml:
```diff
@@ -5,7 +5,7 @@ homepage: https://www.github.com/dart-lang/characters
 environment:
   # This must remain a tight constraint until nnbd is stable
-  sdk: '>=2.10.0-78 <2.10.0'
+  sdk: '>=2.10.0-137 <2.10.0'
```
Note that this might be past the 2.10 stable branch point (and never included in 2.10)

jakemac53 replied:
See https://github.com/dart-lang/sdk/issues/43485, also please mention there that you also were mislead by this

lrhn replied:
I wasn't actually misled, I was just guessing wildly at why Travis gives errors in the sourcespan package, so I updated to the same SDK version as that package. Didn't work, will revert.

lrhn committed:
A version too far.
